packages/mattermost-redux/src/utils/post_list.ts

```diff
@@ -323,142 +322,100 @@ export function makeGenerateCombinedPost(): (state: GlobalState, combinedId: str
     if (systemPosts.length === 0) {
         return null;
     }
-
-    const userActivities = systemPosts.reduce((acc: any, post: Post) => {
+    const userActivities: ActivityEntry[] = [];
+    let prevPost: any = {};
+    systemPosts.reverse().forEach((post: Post) => {
         const postType = post.type;
-        let userActivityProps = acc;
-        const combinedPostType = userActivityProps[postType as string];
-
-        if (
-            postType === Posts.POST_TYPES.ADD_TO_TEAM ||
-            postType === Posts.POST_TYPES.ADD_TO_CHANNEL ||
-            postType === Posts.POST_TYPES.REMOVE_FROM_CHANNEL
-        ) {
+        let actorId: string = post.user_id;
+        if (isUsersRelatedPost(postType)) {
+            // sets the propsUserId to an empty string, this allows the function to combine different "remove" messages with different actors.
+            if (postType === Posts.POST_TYPES.REMOVE_FROM_CHANNEL) {
+                actorId = '';
+            }
+            if (!post.user_id || !post.props) {
+                throw new Error(`Invalid post object: ${JSON.stringify(post)}`);
+            }
             const userId = post.props.addedUserId || post.props.removedUserId;
             const username = post.props.addedUsername || post.props.removedUsername;
-            if (combinedPostType) {
-                if (Array.isArray(combinedPostType[post.user_id])) {
-                    throw new Error('Invalid Post activity data');
-                }
-                const users = combinedPostType[post.user_id] || {ids: [], usernames: []};
-                if (userId) {
-                    if (!users.ids.includes(userId)) {
-                        users.ids.push(userId);
-                    }
-                } else if (username && !users.usernames.includes(username)) {
-                    users.usernames.push(username);
-                }
-                combinedPostType[post.user_id] = users;
-            } else {
-                const users = {
-                    ids: [] as string[],
-                    usernames: [] as string[],
-                };
-
-                if (userId) {
-                    users.ids.push(userId);
-                } else if (username) {
-                    users.usernames.push(username);
-                }
-                userActivityProps[postType] = {
-                    [post.user_id]: users,
-                };
-            }
-        } else {
-            const propsUserId = post.user_id;
 
-            if (combinedPostType) {
-                if (!Array.isArray(combinedPostType)) {
-                    throw new Error('Invalid Post activity data');
-                }
-                if (!combinedPostType.includes(propsUserId)) {
-                    userActivityProps[postType] = [...combinedPostType, propsUserId];
+            // if the post has the same type and userId as the last one, combine them
+            if (prevPost.type === post.type && prevPost.user_id === post.user_id) {
```

Comment:
Instead of storing `prevPost` explicitly, could we instead just look at the last entry in `userActivities` and see if its `actorId` and `postType ` match the current post? That'd also fix a bug that this version has where "removed from channel" messages done by two different users aren't grouped

![Screen Shot 2023-03-02 at 6 23 12 PM](https://user-images.githubusercontent.com/3277310/222586217-992b3bd2-b643-43c0-a3b2-e1c1e2723085.png)